[JSC] Move most of Promise operations to C++

Constellation opened the pull request:
#### 7598b7e88cc0f0ba8736fd178da57e71618ed6d4
<pre>
[JSC] Move most of Promise operations to C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=300130">https://bugs.webkit.org/show_bug.cgi?id=300130</a>
<a href="https://rdar.apple.com/problem/161914098">rdar://problem/161914098</a>

Reviewed by Yijia Huang.

This patch is the second step of optimizing our microtasks.

1. We move most of Promise operations to C++. The remaining ones will be
   moved as well, and only handful of them intentionally remains in JS
   because of callback calls etc.
2. We introduce JSFunctionWithFields to handle various promise resolve
   reject function creations efficiently. This has 2 slots for random
   data, and we leverage this to efficiently create resolve / reject
   functions in C++.
3. Microtask operations are defined via InternalMicrotask, and run fully
   in C++. We use integer to indicate this ID, and we invoke corresponding
   operations in C++. We will increase this coverage, and eventually we will
   stop accepting arbitrary JS functions for microtask. Some remaining
   ones (like, self.queueMicrotask) will be handled via InternalMicrotask
   which invokes JS function.
4. Many promise functions, including Promise.prototype.then is now fully
   written in C++ for performance and strong integration into the rest
   of microtask mechanism.
5. VMEntryRecord is extended to keep JSCell* context field. This is set
   when we invoke a microtask, and unwinding can detect and expand a bit
   additional information via that. We replace async stack trace
   mechanism with this.

* JSTests/stress/never-inlined-functions.js: Removed.
* JSTests/stress/promise-withResolvers.js:
(shouldBe):
* JSTests/stress/sampling-profiler-microtasks.js:
(platformSupportsSamplingProfiler.testResults):
* LayoutTests/inspector/console/message-stack-trace-expected.txt:
* LayoutTests/js/dom/promise-rejection-might-stack-overflow-expected.txt:
* LayoutTests/js/dom/promise-rejection-might-stack-overflow.html:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/AsyncDisposableStackPrototype.js:
(disposeAsync):
* Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js:
(linkTimeConstant.awaitValue): Deleted.
* Source/JavaScriptCore/builtins/AsyncIteratorPrototype.js:
(linkTimeConstant.promiseReturnUndefinedOnFulfilled):
(overriddenName.string_appeared_here.asyncDispose):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/DisposableStackPrototype.js:
(linkTimeConstant.getAsyncDisposableMethod):
* Source/JavaScriptCore/builtins/ModuleLoader.js:
(linkTimeConstant.fulfillFetch):
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(withResolvers): Deleted.
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.promiseEmptyOnRejected):
(linkTimeConstant.newHandledRejectedPromise): Deleted.
(linkTimeConstant.triggerPromiseReactions): Deleted.
(linkTimeConstant.resolvePromise): Deleted.
(linkTimeConstant.rejectPromise): Deleted.
(linkTimeConstant.fulfillPromise): Deleted.
(linkTimeConstant.resolvePromiseWithFirstResolvingFunctionCallCheck): Deleted.
(linkTimeConstant.fulfillPromiseWithFirstResolvingFunctionCallCheck): Deleted.
(linkTimeConstant.rejectPromiseWithFirstResolvingFunctionCallCheck): Deleted.
(linkTimeConstant.createResolvingFunctions): Deleted.
(linkTimeConstant.neverInline.promiseReactionJobWithoutPromise): Deleted.
(linkTimeConstant.resolveWithoutPromise): Deleted.
(linkTimeConstant.rejectWithoutPromise): Deleted.
(linkTimeConstant.fulfillWithoutPromise): Deleted.
(linkTimeConstant.resolveWithoutPromiseForAsyncAwait):
(linkTimeConstant.createResolvingFunctionsWithoutPromise): Deleted.
(linkTimeConstant.neverInline.promiseReactionJob):
(linkTimeConstant.promiseResolveThenableJobFast): Deleted.
(linkTimeConstant.promiseResolveThenableJobWithoutPromiseFast): Deleted.
(linkTimeConstant.promiseResolveThenableJob):
(linkTimeConstant.promiseResolveThenableJobWithDerivedPromise): Deleted.
(linkTimeConstant.alwaysInline.performPromiseThen): Deleted.
* Source/JavaScriptCore/builtins/PromisePrototype.js:
(then): Deleted.
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/debugger/Debugger.h:
* Source/JavaScriptCore/debugger/ScriptProfilingScope.h:
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
(Inspector::JSInjectedScriptHost::isPromiseRejectedWithNativeGetterTypeError):
(Inspector::JSInjectedScriptHost::getInternalProperties):
* Source/JavaScriptCore/interpreter/CachedCall.cpp:
(JSC::CachedCall::CachedCall):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
(JSC::Interpreter::getStackTrace):
(JSC::Interpreter::notifyDebuggerOfExceptionToBeThrown):
(JSC::Interpreter::executeProgram):
(JSC::Interpreter::executeBoundCall):
(JSC::Interpreter::executeCallImpl):
(JSC::Interpreter::executeCall):
(JSC::Interpreter::executeConstruct):
(JSC::Interpreter::executeEval):
(JSC::Interpreter::executeModuleProgram):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/interpreter/ProtoCallFrame.h:
* Source/JavaScriptCore/interpreter/ProtoCallFrameInlines.h:
(JSC::ProtoCallFrame::init):
* Source/JavaScriptCore/interpreter/VMEntryRecord.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/CallData.cpp:
(JSC::call):
(JSC::callMicrotask):
* Source/JavaScriptCore/runtime/CallData.h:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSFunctionWithFields.cpp: Added.
(JSC::JSFunctionWithFields::JSFunctionWithFields):
(JSC::JSFunctionWithFields::create):
(JSC::JSFunctionWithFields::visitChildrenImpl):
(JSC::JSFunctionWithFields::createStructure):
* Source/JavaScriptCore/runtime/JSFunctionWithFields.h: Copied from Source/JavaScriptCore/runtime/JSPromisePrototype.h.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
(JSC::JSGlobalObject::installSaneChainWatchpoints):
(JSC::JSGlobalObject::tryInstallTypedArraySpeciesWatchpoint):
(JSC::JSGlobalObject::queueMicrotask):
(JSC::JSGlobalObject::installArraySpeciesWatchpoint): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseThenWatchpointSet):
(JSC::JSGlobalObject::promiseSpeciesWatchpointSet):
(JSC::JSGlobalObject::promisePrototypeChainIsSaneWatchpointSet):
(JSC::JSGlobalObject::promiseSpeciesGetterSetter const):
(JSC::JSGlobalObject::promiseResolveFunction const):
(JSC::JSGlobalObject::internalPromisePrototype const):
(JSC::JSGlobalObject::functionWithFieldsStructure const):
(JSC::JSGlobalObject::promiseCapabilityObjectStructure const):
(JSC::JSGlobalObject::promiseReactionJobFunction const): Deleted.
(JSC::JSGlobalObject::promiseReactionJobWithoutPromiseFunction const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::asyncGeneratorYieldOnRejectedFunction const):
(JSC::JSGlobalObject::promiseEmptyOnFulfilledFunction const):
(JSC::JSGlobalObject::promiseEmptyOnRejectedFunction const):
(JSC::JSGlobalObject::newPromiseCapabilityFunction const): Deleted.
(JSC::JSGlobalObject::resolvePromiseFunction const): Deleted.
(JSC::JSGlobalObject::rejectPromiseFunction const): Deleted.
(JSC::JSGlobalObject::performPromiseThenFunction const): Deleted.
* Source/JavaScriptCore/runtime/JSInternalPromise.cpp:
(JSC::JSInternalPromise::then):
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::dynamicCastToCell):
(JSC::promiseResolveThenableJobFastSlow):
(JSC::promiseResolveThenableJobWithoutPromiseFastSlow):
(JSC::promiseResolveThenableJob):
(JSC::runInternalMirotask):
(JSC::runJSMicrotask):
* Source/JavaScriptCore/runtime/JSMicrotask.h:
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::createNewPromiseCapability):
(JSC::JSPromise::createPromiseCapability):
(JSC::JSPromise::createDeferredData):
(JSC::JSPromise::rejectedPromise):
(JSC::JSPromise::resolve):
(JSC::JSPromise::reject):
(JSC::JSPromise::fulfill):
(JSC::JSPromise::performPromiseThenExported):
(JSC::JSPromise::rejectAsHandled):
(JSC::JSPromise::performPromiseThen):
(JSC::JSPromise::rejectPromise):
(JSC::JSPromise::fulfillPromise):
(JSC::JSPromise::resolvePromise):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSPromise::triggerPromiseReactions):
(JSC::JSPromise::resolveWithoutPromiseForAsyncAwait):
(JSC::JSPromise::resolveWithoutPromise):
(JSC::JSPromise::rejectWithoutPromise):
(JSC::JSPromise::fulfillWithoutPromise):
(JSC::JSPromise::isThenFastAndNonObservable):
(JSC::promiseSpeciesConstructor):
(JSC::createPromiseCapabilityObjectStructure):
(JSC::JSPromise::then):
(JSC::JSPromise::status const): Deleted.
(JSC::JSPromise::result const): Deleted.
(JSC::JSPromise::flags const): Deleted.
(JSC::JSPromise::isHandled const): Deleted.
(JSC::JSPromise::convertCapabilityToDeferredData): Deleted.
(JSC::callFunction): Deleted.
(JSC::JSPromise::markAsHandled): Deleted.
* Source/JavaScriptCore/runtime/JSPromise.h:
(JSC::JSPromise::status const):
(JSC::JSPromise::isHandled const):
(JSC::JSPromise::result const):
(JSC::JSPromise::reactionsOrResult const):
(JSC::JSPromise::setReactionsOrResult):
(JSC::JSPromise::markAsHandled):
(JSC::JSPromise::flags const):
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSPromiseConstructor::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromisePrototype.cpp:
(JSC::promiseSpeciesWatchpointIsValid):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromisePrototype.h:
* Source/JavaScriptCore/runtime/JSPromiseReaction.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION): Deleted.
* Source/JavaScriptCore/runtime/JSPromiseReaction.h:
* Source/JavaScriptCore/runtime/Microtask.h:
* Source/JavaScriptCore/runtime/MicrotaskQueue.h:
(JSC::QueuedTask::arguments const):
* Source/JavaScriptCore/runtime/MicrotaskQueueInlines.h:
(JSC::MicrotaskQueue::performMicrotaskCheckpoint):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::callPromiseRejectionCallback):
(JSC::VM::didExhaustMicrotaskQueue):
(JSC::VM::promiseResolvingFunctionResolveExecutableSlow):
(JSC::VM::promiseResolvingFunctionRejectExecutableSlow):
(JSC::VM::promiseFirstResolvingFunctionResolveExecutableSlow):
(JSC::VM::promiseFirstResolvingFunctionRejectExecutableSlow):
(JSC::VM::promiseResolvingFunctionResolveWithoutPromiseExecutableSlow):
(JSC::VM::promiseResolvingFunctionRejectWithoutPromiseExecutableSlow):
(JSC::VM::promiseCapabilityExecutorExecutableSlow):
(JSC::VM::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::promiseResolvingFunctionResolveExecutable):
(JSC::VM::promiseResolvingFunctionRejectExecutable):
(JSC::VM::promiseFirstResolvingFunctionResolveExecutable):
(JSC::VM::promiseFirstResolvingFunctionRejectExecutable):
(JSC::VM::promiseResolvingFunctionResolveWithoutPromiseExecutable):
(JSC::VM::promiseResolvingFunctionRejectWithoutPromiseExecutable):
(JSC::VM::promiseCapabilityExecutorExecutable):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
(WebCore::IteratorTraits&gt;::next):
(WebCore::IteratorTraits&gt;::runNextSteps):
* Source/WebCore/bindings/js/JSDOMPromise.cpp:
(WebCore::DOMPromise::result const):
(WebCore::DOMPromise::status const):
(WebCore::DOMPromise::markAsHandled):
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::DeferredPromise::markAsHandled const):
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::queueMicrotaskToEventLoop):
* Source/WebCore/dom/Microtasks.cpp:
(WebCore::MicrotaskQueue::runJSMicrotask):
(WebCore::MicrotaskQueue::performMicrotaskCheckpoint):
* Source/WebCore/dom/RejectedPromiseTracker.cpp:
(WebCore::RejectedPromiseTracker::promiseRejected):
(WebCore::RejectedPromiseTracker::reportUnhandledRejections):
(WebCore::RejectedPromiseTracker::reportRejectionHandled):
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::reportUnhandledPromiseRejection):
* Source/WebCore/dom/UserGestureIndicator.cpp:
(WebCore::UserGestureIndicator::currentUserGestureForMainThread):
* Source/WebCore/dom/UserGestureIndicator.h:
* Source/WebCore/inspector/InspectorFrontendAPIDispatcher.cpp:
(WebCore::InspectorFrontendAPIDispatcher::evaluateOrQueueExpression):

Canonical link: <a href="https://commits.webkit.org/301095@main">https://commits.webkit.org/301095@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64567f29930c9401ecbe55580a3d401286441cd4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124927 "13 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44596 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35333 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131776 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cdf51d1e-1ba9-486b-b1ad-cae37df7ee8f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126804 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53163 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/95078 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/847c685f-6737-4a13-8f0b-8e406cb7e68c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127881 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36138 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111724 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75627 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c26ee7ad-1841-4ccd-9f08-fbc31a74aef4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35073 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29878 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75254 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117028 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105899 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30109 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134448 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123446 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51758 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39557 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/103555 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52179 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107938 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103329 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48670 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26961 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48777 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19577 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51637 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/57439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/156471 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51019 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39188 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54374 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52712 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on JSTests/stress/promise-withResolvers.js:
```diff
@@ -22,11 +22,11 @@ shouldBe(Promise.withResolvers.length, 0);
 
     shouldBe(resolve instanceof Function, true);
-    shouldBe(resolve.name, "resolve");
+    shouldBe(resolve.name, "");
```
This is correct since https://tc39.es/ecma262/#sec-promise-resolve-functions: `A promise resolve function is an anonymous built-in function ...`. So, ECMAScript doesn't require these functions to have specific names.

Constellation replied:
Yes, this is improvement.

sosukesuzuki replied:
FYI I updated test262 after seeing this change: https://github.com/tc39/test262/pull/4594

hyjorc1 commented on JSTests/stress/promise-withResolvers.js:
```diff
@@ -22,11 +22,11 @@ shouldBe(Promise.withResolvers.length, 0);
 
     shouldBe(reject instanceof Function, true);
-    shouldBe(reject.name, "reject");
+    shouldBe(reject.name, "");
```
This is correct since https://tc39.es/ecma262/#sec-promise-reject-functions: `A promise reject function is an anonymous built-in function ...`. So, ECMAScript doesn't require these functions to have specific names.

Constellation replied:
Right. This is improvement.

hyjorc1 commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -261,20 +262,520 @@ JSPromise* JSPromise::rejectWithCaughtException(JSGlobalObject* globalObject, Th
+        return JSValue::encode(jsUndefined());
+
+    callee->fields()[1].set(vm, callee, jsNull());
+    other->fields()[1].set(vm, callee, jsNull());
```
Should it be `other->fields()[1].set(vm, other, jsNull());`?

Constellation replied:
Oh, right. Fixed.

hyjorc1 commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -261,20 +262,520 @@ JSPromise* JSPromise::rejectWithCaughtException(JSGlobalObject* globalObject, Th
+        return JSValue::encode(jsUndefined());
+
+    callee->fields()[1].set(vm, callee, jsNull());
+    other->fields()[1].set(vm, callee, jsNull());
```
ditto

Constellation replied:
Fixed.

hyjorc1 commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -261,20 +262,520 @@ JSPromise* JSPromise::rejectWithCaughtException(JSGlobalObject* globalObject, Th
+        return JSValue::encode(jsUndefined());
+
+    callee->fields()[1].set(vm, callee, jsNull());
+    other->fields()[1].set(vm, callee, jsNull());
```
ditto

Constellation replied:
Fixed.

hyjorc1 commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -261,20 +262,520 @@ JSPromise* JSPromise::rejectWithCaughtException(JSGlobalObject* globalObject, Th
+        return JSValue::encode(jsUndefined());
+
+    callee->fields()[1].set(vm, callee, jsNull());
+    other->fields()[1].set(vm, callee, jsNull());
```
ditto

Constellation replied:
Fixed.

hyjorc1 reviewed (approved):
Nice! r=me!

kmiller68 commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -78,70 +87,81 @@ void JSPromise::visitChildrenImpl(JSCell* cell, Visitor& visitor)
-    auto scope = DECLARE_THROW_SCOPE(vm);
+    auto* executor = JSFunctionWithFields::create(vm, globalObject, vm.promiseCapabilityExecutorExecutable(), 2, nullString());
+    executor->fields()[0].set(vm, executor, jsUndefined());
+    executor->fields()[1].set(vm, executor, jsUndefined());
```
We should really have these be named in some way. It's consistently been a long term matainance burden when these are just offsets and not named fields. Even just having an enum we consistently use would be fine.

Constellation replied:
Done

kmiller68 commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -261,20 +262,519 @@ JSPromise* JSPromise::rejectWithCaughtException(JSGlobalObject* globalObject, Th
+    if (!reject.isUndefined()) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "reject function is already set"_s);
+
+    callee->setField(vm, JSFunctionWithFields::Field::ExecutorResolve, callFrame->argument(0));
+    callee->setField(vm, JSFunctionWithFields::Field::ExecutorReject, callFrame->argument(1));
```
Cool much clearer, thanks.

Constellation committed:
[JSC] Move most of Promise operations to C++
https://bugs.webkit.org/show_bug.cgi?id=300130
rdar://problem/161914098

Reviewed by Yijia Huang.

This patch is the second step of optimizing our microtasks.

1. We move most of Promise operations to C++. The remaining ones will be
   moved as well, and only handful of them intentionally remains in JS
   because of callback calls etc.
2. We introduce JSFunctionWithFields to handle various promise resolve
   reject function creations efficiently. This has 2 slots for random
   data, and we leverage this to efficiently create resolve / reject
   functions in C++.
3. Microtask operations are defined via InternalMicrotask, and run fully
   in C++. We use integer to indicate this ID, and we invoke corresponding
   operations in C++. We will increase this coverage, and eventually we will
   stop accepting arbitrary JS functions for microtask. Some remaining
   ones (like, self.queueMicrotask) will be handled via InternalMicrotask
   which invokes JS function.
4. Many promise functions, including Promise.prototype.then is now fully
   written in C++ for performance and strong integration into the rest
   of microtask mechanism.
5. VMEntryRecord is extended to keep JSCell* context field. This is set
   when we invoke a microtask, and unwinding can detect and expand a bit
   additional information via that. We replace async stack trace
   mechanism with this.

* JSTests/stress/never-inlined-functions.js: Removed.
* JSTests/stress/promise-withResolvers.js:
(shouldBe):
* JSTests/stress/sampling-profiler-microtasks.js:
(platformSupportsSamplingProfiler.testResults):
* LayoutTests/inspector/console/message-stack-trace-expected.txt:
* LayoutTests/js/dom/promise-rejection-might-stack-overflow-expected.txt:
* LayoutTests/js/dom/promise-rejection-might-stack-overflow.html:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/AsyncDisposableStackPrototype.js:
(disposeAsync):
* Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js:
(linkTimeConstant.awaitValue): Deleted.
* Source/JavaScriptCore/builtins/AsyncIteratorPrototype.js:
(linkTimeConstant.promiseReturnUndefinedOnFulfilled):
(overriddenName.string_appeared_here.asyncDispose):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/DisposableStackPrototype.js:
(linkTimeConstant.getAsyncDisposableMethod):
* Source/JavaScriptCore/builtins/ModuleLoader.js:
(linkTimeConstant.fulfillFetch):
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(withResolvers): Deleted.
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.promiseEmptyOnRejected):
(linkTimeConstant.newHandledRejectedPromise): Deleted.
(linkTimeConstant.triggerPromiseReactions): Deleted.
(linkTimeConstant.resolvePromise): Deleted.
(linkTimeConstant.rejectPromise): Deleted.
(linkTimeConstant.fulfillPromise): Deleted.
(linkTimeConstant.resolvePromiseWithFirstResolvingFunctionCallCheck): Deleted.
(linkTimeConstant.fulfillPromiseWithFirstResolvingFunctionCallCheck): Deleted.
(linkTimeConstant.rejectPromiseWithFirstResolvingFunctionCallCheck): Deleted.
(linkTimeConstant.createResolvingFunctions): Deleted.
(linkTimeConstant.neverInline.promiseReactionJobWithoutPromise): Deleted.
(linkTimeConstant.resolveWithoutPromise): Deleted.
(linkTimeConstant.rejectWithoutPromise): Deleted.
(linkTimeConstant.fulfillWithoutPromise): Deleted.
(linkTimeConstant.resolveWithoutPromiseForAsyncAwait):
(linkTimeConstant.createResolvingFunctionsWithoutPromise): Deleted.
(linkTimeConstant.neverInline.promiseReactionJob):
(linkTimeConstant.promiseResolveThenableJobFast): Deleted.
(linkTimeConstant.promiseResolveThenableJobWithoutPromiseFast): Deleted.
(linkTimeConstant.promiseResolveThenableJob):
(linkTimeConstant.promiseResolveThenableJobWithDerivedPromise): Deleted.
(linkTimeConstant.alwaysInline.performPromiseThen): Deleted.
* Source/JavaScriptCore/builtins/PromisePrototype.js:
(then): Deleted.
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/debugger/Debugger.h:
* Source/JavaScriptCore/debugger/ScriptProfilingScope.h:
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/inspector/JSInjectedScriptHost.cpp:
(Inspector::JSInjectedScriptHost::isPromiseRejectedWithNativeGetterTypeError):
(Inspector::JSInjectedScriptHost::getInternalProperties):
* Source/JavaScriptCore/interpreter/CachedCall.cpp:
(JSC::CachedCall::CachedCall):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
(JSC::Interpreter::getStackTrace):
(JSC::Interpreter::notifyDebuggerOfExceptionToBeThrown):
(JSC::Interpreter::executeProgram):
(JSC::Interpreter::executeBoundCall):
(JSC::Interpreter::executeCallImpl):
(JSC::Interpreter::executeCall):
(JSC::Interpreter::executeConstruct):
(JSC::Interpreter::executeEval):
(JSC::Interpreter::executeModuleProgram):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/interpreter/ProtoCallFrame.h:
* Source/JavaScriptCore/interpreter/ProtoCallFrameInlines.h:
(JSC::ProtoCallFrame::init):
* Source/JavaScriptCore/interpreter/VMEntryRecord.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/CallData.cpp:
(JSC::call):
(JSC::callMicrotask):
* Source/JavaScriptCore/runtime/CallData.h:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSFunctionWithFields.cpp: Added.
(JSC::JSFunctionWithFields::JSFunctionWithFields):
(JSC::JSFunctionWithFields::create):
(JSC::JSFunctionWithFields::visitChildrenImpl):
(JSC::JSFunctionWithFields::createStructure):
* Source/JavaScriptCore/runtime/JSFunctionWithFields.h: Copied from Source/JavaScriptCore/runtime/JSPromisePrototype.h.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
(JSC::JSGlobalObject::installSaneChainWatchpoints):
(JSC::JSGlobalObject::tryInstallTypedArraySpeciesWatchpoint):
(JSC::JSGlobalObject::queueMicrotask):
(JSC::JSGlobalObject::installArraySpeciesWatchpoint): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseThenWatchpointSet):
(JSC::JSGlobalObject::promiseSpeciesWatchpointSet):
(JSC::JSGlobalObject::promisePrototypeChainIsSaneWatchpointSet):
(JSC::JSGlobalObject::promiseSpeciesGetterSetter const):
(JSC::JSGlobalObject::promiseResolveFunction const):
(JSC::JSGlobalObject::internalPromisePrototype const):
(JSC::JSGlobalObject::functionWithFieldsStructure const):
(JSC::JSGlobalObject::promiseCapabilityObjectStructure const):
(JSC::JSGlobalObject::promiseReactionJobFunction const): Deleted.
(JSC::JSGlobalObject::promiseReactionJobWithoutPromiseFunction const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::asyncGeneratorYieldOnRejectedFunction const):
(JSC::JSGlobalObject::promiseEmptyOnFulfilledFunction const):
(JSC::JSGlobalObject::promiseEmptyOnRejectedFunction const):
(JSC::JSGlobalObject::newPromiseCapabilityFunction const): Deleted.
(JSC::JSGlobalObject::resolvePromiseFunction const): Deleted.
(JSC::JSGlobalObject::rejectPromiseFunction const): Deleted.
(JSC::JSGlobalObject::performPromiseThenFunction const): Deleted.
* Source/JavaScriptCore/runtime/JSInternalPromise.cpp:
(JSC::JSInternalPromise::then):
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::dynamicCastToCell):
(JSC::promiseResolveThenableJobFastSlow):
(JSC::promiseResolveThenableJobWithoutPromiseFastSlow):
(JSC::promiseResolveThenableJob):
(JSC::runInternalMirotask):
(JSC::runJSMicrotask):
* Source/JavaScriptCore/runtime/JSMicrotask.h:
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::createNewPromiseCapability):
(JSC::JSPromise::createPromiseCapability):
(JSC::JSPromise::createDeferredData):
(JSC::JSPromise::rejectedPromise):
(JSC::JSPromise::resolve):
(JSC::JSPromise::reject):
(JSC::JSPromise::fulfill):
(JSC::JSPromise::performPromiseThenExported):
(JSC::JSPromise::rejectAsHandled):
(JSC::JSPromise::performPromiseThen):
(JSC::JSPromise::rejectPromise):
(JSC::JSPromise::fulfillPromise):
(JSC::JSPromise::resolvePromise):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSPromise::triggerPromiseReactions):
(JSC::JSPromise::resolveWithoutPromiseForAsyncAwait):
(JSC::JSPromise::resolveWithoutPromise):
(JSC::JSPromise::rejectWithoutPromise):
(JSC::JSPromise::fulfillWithoutPromise):
(JSC::JSPromise::isThenFastAndNonObservable):
(JSC::promiseSpeciesConstructor):
(JSC::createPromiseCapabilityObjectStructure):
(JSC::JSPromise::then):
(JSC::JSPromise::status const): Deleted.
(JSC::JSPromise::result const): Deleted.
(JSC::JSPromise::flags const): Deleted.
(JSC::JSPromise::isHandled const): Deleted.
(JSC::JSPromise::convertCapabilityToDeferredData): Deleted.
(JSC::callFunction): Deleted.
(JSC::JSPromise::markAsHandled): Deleted.
* Source/JavaScriptCore/runtime/JSPromise.h:
(JSC::JSPromise::status const):
(JSC::JSPromise::isHandled const):
(JSC::JSPromise::result const):
(JSC::JSPromise::reactionsOrResult const):
(JSC::JSPromise::setReactionsOrResult):
(JSC::JSPromise::markAsHandled):
(JSC::JSPromise::flags const):
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSPromiseConstructor::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromisePrototype.cpp:
(JSC::promiseSpeciesWatchpointIsValid):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromisePrototype.h:
* Source/JavaScriptCore/runtime/JSPromiseReaction.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION): Deleted.
* Source/JavaScriptCore/runtime/JSPromiseReaction.h:
* Source/JavaScriptCore/runtime/Microtask.h:
* Source/JavaScriptCore/runtime/MicrotaskQueue.h:
(JSC::QueuedTask::arguments const):
* Source/JavaScriptCore/runtime/MicrotaskQueueInlines.h:
(JSC::MicrotaskQueue::performMicrotaskCheckpoint):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::callPromiseRejectionCallback):
(JSC::VM::didExhaustMicrotaskQueue):
(JSC::VM::promiseResolvingFunctionResolveExecutableSlow):
(JSC::VM::promiseResolvingFunctionRejectExecutableSlow):
(JSC::VM::promiseFirstResolvingFunctionResolveExecutableSlow):
(JSC::VM::promiseFirstResolvingFunctionRejectExecutableSlow):
(JSC::VM::promiseResolvingFunctionResolveWithoutPromiseExecutableSlow):
(JSC::VM::promiseResolvingFunctionRejectWithoutPromiseExecutableSlow):
(JSC::VM::promiseCapabilityExecutorExecutableSlow):
(JSC::VM::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::promiseResolvingFunctionResolveExecutable):
(JSC::VM::promiseResolvingFunctionRejectExecutable):
(JSC::VM::promiseFirstResolvingFunctionResolveExecutable):
(JSC::VM::promiseFirstResolvingFunctionRejectExecutable):
(JSC::VM::promiseResolvingFunctionResolveWithoutPromiseExecutable):
(JSC::VM::promiseResolvingFunctionRejectWithoutPromiseExecutable):
(JSC::VM::promiseCapabilityExecutorExecutable):
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
(WebCore::IteratorTraits>::next):
(WebCore::IteratorTraits>::runNextSteps):
* Source/WebCore/bindings/js/JSDOMPromise.cpp:
(WebCore::DOMPromise::result const):
(WebCore::DOMPromise::status const):
(WebCore::DOMPromise::markAsHandled):
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::DeferredPromise::markAsHandled const):
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::queueMicrotaskToEventLoop):
* Source/WebCore/dom/Microtasks.cpp:
(WebCore::MicrotaskQueue::runJSMicrotask):
(WebCore::MicrotaskQueue::performMicrotaskCheckpoint):
* Source/WebCore/dom/RejectedPromiseTracker.cpp:
(WebCore::RejectedPromiseTracker::promiseRejected):
(WebCore::RejectedPromiseTracker::reportUnhandledRejections):
(WebCore::RejectedPromiseTracker::reportRejectionHandled):
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::reportUnhandledPromiseRejection):
* Source/WebCore/dom/UserGestureIndicator.cpp:
(WebCore::UserGestureIndicator::currentUserGestureForMainThread):
* Source/WebCore/dom/UserGestureIndicator.h:
* Source/WebCore/inspector/InspectorFrontendAPIDispatcher.cpp:
(WebCore::InspectorFrontendAPIDispatcher::evaluateOrQueueExpression):

Canonical link: https://commits.webkit.org/301095@main
